nixos/modules/virtualisation/hyperv-guest.nix

```
@@ -25,40 +28,69 @@ in {
     };
   };
 
-  config = mkIf cfg.enable {
-    boot = {
-      initrd.kernelModules = [
-        "hv_balloon" "hv_netvsc" "hv_storvsc" "hv_utils" "hv_vmbus"
-      ];
+  config = mkIf cfg.enable (mkMerge [
+    {
+      boot = {
+        initrd.kernelModules = [
+          "hv_balloon" "hv_netvsc" "hv_storvsc" "hv_utils" "hv_vmbus"
+        ];
 
-      initrd.availableKernelModules = [ "hyperv_keyboard" ];
+        initrd.availableKernelModules = [ "hyperv_keyboard" ];
 
-      kernelParams = [
-        "video=hyperv_fb:${cfg.videoMode}" "elevator=noop"
-      ];
-    };
+        kernelParams = [
+          "video=hyperv_fb:${cfg.videoMode}" "elevator=noop"
+        ];
+        blacklistedKernelModules = [ "hyperv_fb" ]; # workaround to use hyperv_drm
+      };
 
-    environment.systemPackages = [ config.boot.kernelPackages.hyperv-daemons.bin ];
+      environment.systemPackages = [ config.boot.kernelPackages.hyperv-daemons.bin ];
 
-    # enable hotadding cpu/memory
-    services.udev.packages = lib.singleton (pkgs.writeTextFile {
-      name = "hyperv-cpu-and-memory-hotadd-udev-rules";
-      destination = "/etc/udev/rules.d/99-hyperv-cpu-and-memory-hotadd.rules";
-      text = ''
-        # Memory hotadd
-        SUBSYSTEM=="memory", ACTION=="add", DEVPATH=="/devices/system/memory/memory[0-9]*", TEST=="state", ATTR{state}="online"
+      # enable hotadding cpu/memory
+      services.udev.packages = lib.singleton (pkgs.writeTextFile {
+        name = "hyperv-cpu-and-memory-hotadd-udev-rules";
+        destination = "/etc/udev/rules.d/99-hyperv-cpu-and-memory-hotadd.rules";
+        text = ''
+          # Memory hotadd
+          SUBSYSTEM=="memory", ACTION=="add", DEVPATH=="/devices/system/memory/memory[0-9]*", TEST=="state", ATTR{state}="online"
 
-        # CPU hotadd
-        SUBSYSTEM=="cpu", ACTION=="add", DEVPATH=="/devices/system/cpu/cpu[0-9]*", TEST=="online", ATTR{online}="1"
-      '';
-    });
+          # CPU hotadd
+          SUBSYSTEM=="cpu", ACTION=="add", DEVPATH=="/devices/system/cpu/cpu[0-9]*", TEST=="online", ATTR{online}="1"
+        '';
+      });
 
-    systemd = {
-      packages = [ config.boot.kernelPackages.hyperv-daemons.lib ];
+      systemd = {
+        packages = [ config.boot.kernelPackages.hyperv-daemons.lib ];
 
-      targets.hyperv-daemons = {
-        wantedBy = [ "multi-user.target" ];
+        targets.hyperv-daemons = {
+          wantedBy = [ "multi-user.target" ];
+        };
       };
-    };
-  };
+    }
+    (mkIf cfg.enable {
+      assertions = [{
+        assertion = config.hardware.pulseaudio.enable;
+        message = ''
+          Audio support for Hyper-V (xrdp) currently only supports pulseaudio
+        '';
+      }];
+      services.xrdp = {
+        enable = true;
+        openFirewall = true;
+        transportMode = "vsock";
+        address = "-1";
```

```suggestion
        address = "-1"; # VMADDR_CID_ANY
```
It may be helpful for future readers to know what a -1 address is.